File: src/android/OcrPlugin.java
package jp.bluememe.plugin.ocr;

import android.content.Context;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import jp.bluememe.plugin.ocr.plugin.OcrPluginManager;

public class OcrPlugin extends CordovaPlugin {

    // MARK: - Callback
    private CallbackContext mOcrCallbackId = null;

    // MARK: - Member
    private OcrPluginManager mOcrPluginManager = null;

    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) throws JSONException {
        Context applicationContext = cordova.getActivity().getApplicationContext();
        if (action.equals("startOCR")) {
            mOcrCallbackId = callbackContext;
            startOcr();
            return true;
        }
        return false;
    }

    // MARK: - Function
    private OcrPluginManager getOcrPluginManager() {
        if (mOcrPluginManager == null) {
            mOcrPluginManager = new OcrPluginManager();
        }
        return mOcrPluginManager;
    }

    private void startOcr() {
        getOcrPluginManager().start(cordova.getActivity(), new OcrPluginManager.OnOcrPluginManagerListener() {
            @Override
            public void onResult(OcrPluginManager sender, int code, OcrPluginManager.OcrResultInfo result) {
                ocrStop();
                JSONObject resultJson = new JSONObject();
                if (code == OcrPluginManager.CODE_SUCCESS) {
                    if (result == null) {
                        try {
                            resultJson.put("errorCode", -1);
                        } catch (JSONException e) {
                        }
                        mOcrCallbackId.error(resultJson);
                    } else {
                        try {
                            resultJson.put("errorCode", 0);
                            resultJson.put("type", result.mCardType);
                            resultJson.put("name", result.mName);
                            resultJson.put("address", result.mAddress);
                            resultJson.put("birthdate", result.mBirthdate);
                            resultJson.put("gender", result.mGender);
                        } catch (JSONException e) {
                        }
                        mOcrCallbackId.error(resultJson);
                    }
                } else if (code == OcrPluginManager.CODE_AUTHROIZE) {
                    try {
                        resultJson.put("errorCode", -2);
                    } catch (JSONException e) {
                    }
                    mOcrCallbackId.error(resultJson);
                } else {
                    try {
                        resultJson.put("errorCode", -1);
                    } catch (JSONException e) {
                    }
                    mOcrCallbackId.error(resultJson);
                }
            }
        });
    }

    private void ocrStop() {
        getOcrPluginManager().stop();
        mOcrPluginManager = null;
    }
}

File: src/ios/OcrPlugin.swift
//
//  OcrPlugin.swift
//  CardReaderSample
//
//  Created by 平山 裕也 on 2023/12/04.
//

import Foundation
import UIKit

@objc(OcrPlugin)
class OcrPlugin : CDVPlugin, OnOcrPluginManagerDelegate {
    
    // MARK: - Member
    private var mOcrPluginManager: OcrPluginManager? = nil
    private var mOcrCallbackId: String? = nil
    
    // MARK: - OnOcrPluginManagerDelegate
    func onResult(sender: OcrPluginManager, code: Int, result: OcrResultInfo?) {
        stopOcr()
        if (mOcrCallbackId != nil) {
            if (code == OcrPluginManager.CODE_SUCCESS) {
                let inType = result?.mCardType ?? 0
                let inName = result?.mName ?? ""
                let inGender = result?.mGender ?? ""
                let inAddress = result?.mAddress ?? ""
                let inBirthdate = result?.mBirthdate ?? ""
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: ["errorCode": 0, "type": inType, "name": inName, "gender": inGender, "address": inAddress, "birthdate": inBirthdate ])
                pluginResult?.keepCallback = true
                commandDelegate.send(pluginResult, callbackId: mOcrCallbackId)
            } else if (code == OcrPluginManager.CODE_AUTHROIZE) {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: ["errorCode": -2] )
                pluginResult?.keepCallback = true
                commandDelegate.send(pluginResult, callbackId: mOcrCallbackId)
            } else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: ["errorCode": -1] )
                pluginResult?.keepCallback = true
                commandDelegate.send(pluginResult, callbackId: mOcrCallbackId)
            }
        }
    }
        
    // MARK: - Accesser
    @objc(startOCR:)
    func startOCR(command: CDVInvokedUrlCommand) {
        mOcrCallbackId = command.callbackId
        startOcr()
    }
    
    // MARK: - Function
    private func getOcrPluginManager() -> OcrPluginManager {
        if (mOcrPluginManager == nil) {
            mOcrPluginManager = OcrPluginManager()
        }
        return mOcrPluginManager!
    }
    
    private func startOcr() {
        if (viewController == nil) {
            if (mOcrCallbackId != nil) {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: ["errorCode": -1] )
                pluginResult?.keepCallback = true
                commandDelegate.send(pluginResult, callbackId: mOcrCallbackId)
            }
            return
        }
        let parent = OcrViewController()
        getOcrPluginManager().start(parent: parent, delegate: self)
        viewController.present(parent, animated: true)
        
//        getOcrPluginManager().start(parent: viewController!, delegate: self)
    }
    
    private func stopOcr() {
        getOcrPluginManager().stop()
        mOcrPluginManager = nil
    }

}
